---
title: "GSEA figures"
author: "Mikhail Dozmorov"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: no
  html_document:
    theme: cerulean
    toc: yes
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set up the environment
library(knitr)
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(dplyr)
options(stringsAsFactors = FALSE)
```

```{r libraries}
# Libraries

library(readxl)
library(cowplot)
library(ggplot2)
library(stringr)
```

```{r settings}
# Settings

# File name with the results of GSEA analysis
data_dir       <- "/Users/mdozmorov/Documents/Data/VCU_work/Eun_Lee/2020-12.EunLee_Novogene/" # Working directory
fileName   <- paste0(data_dir, "results/GSEA_DEGs.xlsx") # File to save all results of differential expression analysis
max_GO_to_plot = 12   # Maximum number of GOs to plot
max_enrichment_length <- 50 # Maximum length of enrichment descriptions
```

# Visualization

- Two types of Enrichment analyses were run:
    - **Enrich** - enrichment _without considering_ fold change of DEGs
    - **GSEA** - enrichment _considering_ the order of up- and downregulated DEGs. **NES** - normalized enrichment score - if positive, the process is enriched in UPregulate genes, and vice versa.

- The following signatures were used:
    - **KEGG** - KEGG pathways
    - MSigDB collections, see  https://www.gsea-msigdb.org/gsea/msigdb/index.jsp. 
        - **H**	hallmark gene sets  are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
        - **C1**	positional gene sets  for each human chromosome and cytogenetic band.
        - **C2**	curated gene sets  from online pathway databases, publications in PubMed, and knowledge of domain experts.
        - **C3**	regulatory target gene sets  based on gene target predictions for microRNA seed sequences and predicted transcription factor binding sites.
        - **C4**	computational gene sets  defined by mining large collections of cancer-oriented microarray data.
        - **C5**	GO gene sets  consist of genes annotated by the same GO terms.
        - **C6**	oncogenic gene sets  defined directly from microarray gene expression data from cancer gene perturbations.
        - **C7**	immunologic gene sets  defined directly from microarray gene expression data from immunologic studies.
- Sheets in the Excel results file are named by the **type of analysis**.**signature**, e.g., **Enrich.KEGG**, **GSEA.C7**

**Signatures in question can be googled**

## -log10 adjusted p-value interpretation

$$1 = 10^{-1} = 0.1$$
$$1.3 = 10^{-1.3} \approx 0.05$$
$$2 = 10^{-2} = 0.01$$
$$4 = 10^{-4} = 0.0001$$
etc. -log10 adjusted p-value $\ge1$ is acceptable

```{r fig.height=4}
# all_sheets <- excel_sheets(fileName)
# Rearrange sheets
all_sheets <- c("Enrich.KEGG", "GSEA.KEGG", "Enrich.C5", "GSEA.C5", "Enrich.C7", "GSEA.C7")

# Go through each sheet
for (sheet in all_sheets) {
  # Analysis type
  print(sheet)
  # Process Enrichr results
  if (sheet == "Enrich.KEGG") {
    mtx <- read_xlsx(fileName, sheet = sheet)
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add -log10 adjusted p-value
    mtx$log10padj <- -log10(as.numeric(mtx$Adjusted.P.value))
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Term, width = max_enrichment_length)
  }
  # Process GSEA results
  if (grepl("GSEA", sheet)) {
    mtx <- read_xlsx(fileName, sheet = sheet)
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add -log10 adjusted p-value
    mtx$log10padj <- -log10(as.numeric(mtx$p.adjust))
    # Change the direction, if negative enrichment score
    mtx$log10padj <- mtx$log10padj * ifelse(mtx$NES > 0, 1, -1)
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Description, width = max_enrichment_length)
  }
  # Process MSigDb data
  if (grepl("Enrich.C", sheet)) {
    mtx <- read_xlsx(fileName, sheet = sheet)
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add -log10 adjusted p-value
    mtx$log10padj <- -log10(as.numeric(mtx$p.adjust))
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Description, width = max_enrichment_length)
  }
  
  # Actual plotting
  
  p <- ggplot(data = mtx, aes(x = reorder(enrichment, log10padj), y = log10padj)) +
    geom_col(aes(fill = log10padj)) +
    scale_fill_gradient2(low = "blue", high = "red", midpoint = 0) +
    coord_flip() +
    theme_cowplot() +
    ylab("-log10 adjusted p-value") +
    xlab("Enrichment term")
  print(p)
}


```


